### PR TITLE
Batch integration refactoring

### PR DESCRIPTION
**Pull request by rcannood**
An updated version of #6 which adds several batch integration components to v2.

Future work:
* Implement the benchmarking pipeline
* Migrate more components from the v1 pipeline

I added these items to the project board → https://github.com/orgs/openproblems-bio/projects/9/views/1

### CODE REVIEW

File: src/batch_integration/embedding/methods/combat/script.py [hunk 1]
```diff
@@ -1,54 +1,65 @@
...
-    print('Select HVGs')
+    print('Select HVGs', flush=True)
     adata = adata[:, adata.var['highly_variable']].copy()
 
 if scaling:
-    print('Scale')
+    print('Scale', flush=True)
     adata.X = adata.layers['logcounts_scaled']
 else:
     adata.X = adata.layers['logcounts']
```
Comment (KaiWaldrant) [inline src/batch_integration/embedding/methods/combat/script.py:34]: change to new normalized layers

File: src/batch_integration/embedding/methods/combat/script.py [hunk 2]
```diff
@@ -1,54 +1,65 @@
...
+output = sc.AnnData(
+    obs= adata.obs[[]],
+    obsm={
+        'X_emb': X_emb
+    },
+    uns={
+        'dataset_id': adata.uns['dataset_id'],
+        'normalization_id': adata.uns['normalization_id'],
+        'method_id': meta['functionality_name'],
+        'scaled': par['scaling'],
```
Comment (KaiWaldrant) [inline src/batch_integration/embedding/methods/combat/script.py:58]: remove scaling

File: src/batch_integration/embedding/methods/combat/script.py [hunk 3]
```diff
@@ -1,54 +1,65 @@
...
 ## VIASH START
 par = {
-    'input': './src/batch_integration/resources/datasets_pancreas.h5ad',
-    'output': './src/batch_integration/resources/pancreas_bbknn.h5ad',
+    'input': 'resources_test/batch_integration/pancreas/processed.h5ad',
+    'output': 'output.h5ad',
     'hvg': True,
-    'scaling': True,
-    'debug': True
+    'scaling': True
```
Comment (KaiWaldrant) [inline src/batch_integration/embedding/methods/combat/script.py:9]: remove scaling, incorperated in normalization

File: src/batch_integration/embedding/methods/combat/script.py (hunk 2 above)
Comment (KaiWaldrant) [inline src/batch_integration/embedding/methods/combat/script.py:58]: remove scaling, see above

File: src/batch_integration/embedding/methods/combat/script.py [hunk 4]
```diff
@@ -1,54 +1,65 @@
...
+        'method_id': meta['functionality_name'],
+        'scaled': par['scaling'],
+        'hvg': par('hvg')
+    },
+    layers = adata.layers
+)
 
-adata.write(output, compression='gzip')
+print('Write to output', flush=True)
+adata.write_h5ad(par['output'], compression='gzip')
```
Comment (KaiWaldrant) [inline src/batch_integration/embedding/methods/combat/script.py:65]:
```suggestion
input.write_h5ad(par['output'], compression='gzip')
```
change adata to input

**Comment by KaiWaldrant** [comment]
@rcannood split_dataset  and a metric fail on the organism field because I think that is not in the dataset.

File: src/batch_integration/methods/bbknn/config.vsh.yaml [hunk 5]
```diff
@@ -0,0 +1,33 @@
+# use method api spec
+__merge__: ../../api/comp_method_graph.yaml
+functionality:
+  name: bbknn
+  description: "BBKNN: fast batch alignment of single cell transcriptomes"
+  info:
+    method_name: BBKNN
```
[the remark this answers is not in the document]
Comment (KaiWaldrant) [reply src/batch_integration/methods/bbknn/config.vsh.yaml:7]: change it later with an issue. still needs to be done for all tasks

File: src/batch_integration/api/anndata_score.yaml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+      - type: string
+        name: parent_method_id
+        description: previous method id if anndata is from a previous method
+        required: false
+      - type: boolean
+        name: hvg
+        description: If the method was done on hvg or full
+        required: true
+      - type: string
+        name: output_id
```
Comment (KaiWaldrant) [inline src/batch_integration/api/anndata_score.yaml:39]:
```suggestion
        name: output_type
```

File: src/batch_integration/api/anndata_integrated_graph.yaml [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+      - type: string
+        name: parent_method_id
+        description: if anndata passed from a previous method
+        required: false
+      - type: boolean
+        name: hvg
+        description: If the method was done on hvg or full
+        required: true
+      - type: string
+        name: output_id
```
Comment (KaiWaldrant) [inline src/batch_integration/api/anndata_integrated_graph.yaml:28]:
```suggestion
        name: output_type
```

File: src/batch_integration/api/anndata_integrated_feature.yaml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+      - type: string
+        name: parent_method_id
+        description: if anndata passed from a previous method
+        required: false
+      - type: boolean
+        name: hvg
+        description: If the method was done on hvg or full
+        required: true
+      - type: string
+        name: output_id
```
Comment (KaiWaldrant) [inline src/batch_integration/api/anndata_integrated_feature.yaml:28]:
```suggestion
        name: output_type
```

File: src/batch_integration/api/anndata_integrated_embedding.yaml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+      - type: string
+        name: parent_method_id
+        description: if anndata passed from a previous method
+        required: false
+      - type: boolean
+        name: hvg
+        description: If the method was done on hvg or full
+        required: true
+      - type: string
+        name: output_id
```
Comment (KaiWaldrant) [inline src/batch_integration/api/anndata_integrated_embedding.yaml:28]:
```suggestion
        name: output_type
```

File: src/batch_integration/api/anndata_integrated_embedding.yaml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+        required: true
+    uns:
+      - type: string
+        name: method_id
+        description: "A unique identifier for the method"
+        required: true
+      - type: string
+        name: parent_method_id
+        description: if anndata passed from a previous method
+        required: false
```
Comment (KaiWaldrant) [inline src/batch_integration/api/anndata_integrated_embedding.yaml:22]:
```suggestion
```

File: src/batch_integration/api/anndata_integrated_graph.yaml [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+        required: true
+    uns:
+      - type: string
+        name: method_id
+        description: "A unique identifier for the method"
+        required: true
+      - type: string
+        name: parent_method_id
+        description: if anndata passed from a previous method
+        required: false
```
Comment (KaiWaldrant) [inline src/batch_integration/api/anndata_integrated_graph.yaml:22]:
```suggestion
```

File: src/batch_integration/api/anndata_score.yaml [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+        required: true        
+      - type: double
+        name: metric_values
+        description: "The metric values obtained for the given prediction. Must be of same length as 'metric_ids'."
+        multiple: true
+        required: true
+      - type: string
+        name: parent_method_id
+        description: previous method id if anndata is from a previous method
+        required: false
```
Comment (KaiWaldrant) [inline src/batch_integration/api/anndata_score.yaml:33]:
```suggestion
```

File: src/batch_integration/methods/scanorama_feature/script.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+if par['hvg']:
+    print('Select HVGs', flush=True)
+    adata = adata[:, adata.var['hvg']]
+
+print('Run scanorama', flush=True)
+adata.X = adata.layers['normalized']
+adata.layers['corrected_counts'] = scanorama(adata, batch='batch').X
+
+del(adata.X)
```
Comment (KaiWaldrant) [inline src/batch_integration/methods/scanorama_feature/script.py:33]:
```suggestion
del adata.X
```

File: src/batch_integration/metrics/cell_cycle_conservation/config.vsh.yaml [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+      - metric_id: cell_cycle_conservation
+        metric_name: Cell Cycle Conservation
+        min: 0
+        max: 1
+        maximize: true
+  # arguments:
+  #   - name: --organism
+  #     type: string
+  #     description: Name of organism to compute cell cycle scores on
+  #     required: true
```
Comment (KaiWaldrant) [inline src/batch_integration/metrics/cell_cycle_conservation/config.vsh.yaml:20]: remove if not needed
```suggestion
```

File: src/batch_integration/metrics/clustering_overlap/script.py [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+        "metric_ids": [ "ari", "nmi" ],
+        "metric_values": [ ari_score, nmi_score ],
+        "hvg": input.uns['hvg'],
+        'output_type': input.uns['output_type'],
+    }
+)
+
+if 'parent_method_id' in input.uns:
+    output.uns['parent_method_id'] = input.uns['parent_method_id']
+
```
Comment (KaiWaldrant) [inline src/batch_integration/metrics/clustering_overlap/script.py:50]:
```suggestion
```

File: src/batch_integration/workflows/download.tsv
Comment (KaiWaldrant) [inline src/batch_integration/workflows/download.tsv:1]: remove ?

File: src/batch_integration/workflows/run/config.vsh.yaml
Comment (KaiWaldrant) [inline src/batch_integration/workflows/run/config.vsh.yaml:1]: needs to be updated after merge

File: src/batch_integration/workflows/run/main.nf
Comment (KaiWaldrant) [inline src/batch_integration/workflows/run/main.nf:1]: will be updated after merge

File: src/batch_integration/workflows/run/preprocessing.tsv
Comment (KaiWaldrant) [inline src/batch_integration/workflows/run/preprocessing.tsv:1]: remove /update after merge

File: src/batch_integration/workflows/run/run_nextflow.sh
Comment (KaiWaldrant) [inline src/batch_integration/workflows/run/run_nextflow.sh:1]: will be updated after merge

**Review by KaiWaldrant — CHANGES_REQUESTED** [review]

File: src/batch_integration/api/anndata_integrated_feature.yaml [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+        required: true
+    uns:
+      - type: string
+        name: method_id
+        description: "A unique identifier for the method"
+        required: true
+      - type: string
+        name: parent_method_id
+        description: if anndata passed from a previous method
+        required: false
```
Comment (KaiWaldrant) [inline src/batch_integration/api/anndata_integrated_feature.yaml:22]:
```suggestion
```

**Review by KaiWaldrant — CHANGES_REQUESTED** [review]

File: src/batch_integration/api/comp_split_dataset.yaml [hunk 17]
```diff
@@ -0,0 +1,63 @@
+functionality:
+  info:
+    type: split_dataset
+  arguments:
+    - name: "--input"
+      __merge__: anndata_dataset.yaml
+    - name: "--output"
+      __merge__: anndata_unintegrated.yaml
+      direction: output
+  # test_resources:
```
[the remark this answers is not in the document]
Comment (rcannood) [reply src/batch_integration/api/comp_split_dataset.yaml:10]: Or add it back? Could you look into this?

**Pull request merged**